### PR TITLE
[Common] Added Alignment Requirements for CuBLAS heuristics

### PR DESCRIPTION
**Pull request by phu0ngng**
## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by ksivaman — APPROVED** [review]
LGTM

File: transformer_engine/common/gemm/cublaslt_gemm.cu [hunk 1]
```diff
@@ -34,6 +35,16 @@ cudaDataType_t get_cuda_dtype(const transformer_engine::DType t) {
   }
 }
 
+uint32_t _getAlignment(uintptr_t address) {
+  // alignment are in bytes
+  uint32_t alignment = 256;
+  for (; ; alignment /= 2) {
+    if (!(address % alignment)) {
```
Comment (ptrendx) [inline transformer_engine/common/gemm/cublaslt_gemm.cu:42]:
```suggestion
    if ((address % alignment) == 0) {
```

File: transformer_engine/common/gemm/cublaslt_gemm.cu [hunk 2]
```diff
@@ -260,6 +271,22 @@ void cublas_gemm(const Tensor *inputA,
   NVTE_CHECK_CUBLAS(cublasLtMatmulPreferenceSetAttribute(
           preference, CUBLASLT_MATMUL_PREF_MAX_WORKSPACE_BYTES,
           &workspaceSize, sizeof(workspaceSize)));
+  const auto A_aligned = _getAlignment(reinterpret_cast<uintptr_t>(A));
+  const auto B_aligned = _getAlignment(reinterpret_cast<uintptr_t>(B));
+  const auto C_aligned = _getAlignment(reinterpret_cast<uintptr_t>(C));
+  const auto D_aligned = _getAlignment(reinterpret_cast<uintptr_t>(D));
```
Comment (ptrendx) [inline transformer_engine/common/gemm/cublaslt_gemm.cu:277]:
```suggestion
  const auto A_alignment = _getAlignment(reinterpret_cast<uintptr_t>(A));
  const auto B_alignment = _getAlignment(reinterpret_cast<uintptr_t>(B));
  const auto C_alignment = _getAlignment(reinterpret_cast<uintptr_t>(C));
  const auto D_alignment = _getAlignment(reinterpret_cast<uintptr_t>(D));
```
`A_aligned` for me would indicate a `bool` which says whether A is aligned or not.

**Comment by ptrendx** [comment]
Could we add some test for it?

**Comment by deepakn94** [comment]
LGTM, thanks. I verified that my failing case now works.

**Comment by phu0ngng** [comment]
Hi, thanks for the feedback. Glad to hear that it works for you now @deepakn94.
@ptrendx I added the unit test to capture this bug. If there is no issue, I will merge this PR. Thanks

File: transformer_engine/common/gemm/cublaslt_gemm.cu [hunk 3]
```diff
@@ -34,6 +35,16 @@ cudaDataType_t get_cuda_dtype(const transformer_engine::DType t) {
   }
 }
 
+uint32_t _getAlignment(uintptr_t address) {
```
Comment (ptrendx) [inline transformer_engine/common/gemm/cublaslt_gemm.cu:38]: nit: You don't need to start this function with `_` - it is already with unnamed namespace and not exposed outside this file.

**Review by ptrendx — APPROVED** [review]

**Pull request merged**